lib/ansible/modules/packaging/os/package_facts.py

```diff
@@ -180,14 +180,16 @@ class APT(LibMgr):
 
     @property
     def pkg_cache(self):
-        if self._cache:
+        if self._cache is not None:
             return self._cache
 
         self._cache = self._lib.Cache()
         return self._cache
 
     def list_installed(self):
-        return [pk for pk in self.pkg_cache.keys() if self.pkg_cache[pk].is_installed]
+        # Store the cache to avoid running pkg_cache() for each item in the comprehension, which is very slow
+        cache = self.pkg_cache
+        return [pk for pk in cache.keys() if cache[pk].is_installed]
```

Comment:
This second part shouldn't be needed.  The change to line 183 is sufficient.  The cache is initialized on line 186.  That's the expensive operation that needs to be bypassed which changing the check from checking for truthiness to checking for a sentinel (None) will accomplish.

Reply:
This was a further optimization that did result in a minor but perceptible performance gain.

Reply:
What kind of numbers are we talking about?  function calls are expensive in python but I wouldn't think they're too expensive.  If the additional gain is more than a few seconds (depending on how many packages are installed) I'd be worried that there's something else wrong which would lead to performance problems when using self.pkg_cache in other code.

Reply:
I would have to retest, but I believe it was only a second or two.

Reply:
Cool, maybe in devel, change the comment then.  ```Avoid the function call to the pkg_cache property inside the loop. Saves a couple seconds```  I'll merge this now.

Reply:
Running the code directly on the managed node shows a negligible difference.

With the change that saves the cache to a variable:
```
time python3 AnsiballZ_package_facts.py
...
real	0m1.641s
user	0m1.480s
sys	0m0.156s
```

Without:
```
time python3 ansible-tmp-1567628929.18575-14623375103494-without/AnsiballZ_package_facts.py
...
real	0m1.839s
user	0m1.574s
sys	0m0.260s
```

The time different, probably for many other factors, was more perceptible from the control node:
```
time ansible -i inventory/hosts.yml lab-ub1804 -m package_facts -v # with
...
        4.67 real         1.85 user         0.27 sys
```

```
time ansible -i inventory/hosts.yml lab-ub1804 -m package_facts -v # without
...
        5.41 real         1.93 user         0.28 sys
```